File: app/src/main/java/com/afoxxvi/alopex/dialog/MatchRuleDialog.java
package com.afoxxvi.alopex.dialog;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.Toast;

import com.afoxxvi.alopex.AlopexView;
import com.afoxxvi.alopex.R;
import com.afoxxvi.alopex.databinding.DialogRuleBinding;
import com.afoxxvi.alopex.filter.AlopexFilter;
import com.afoxxvi.alopex.util.Pair;
import com.google.android.material.chip.Chip;
import com.google.android.material.chip.ChipGroup;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

public class MatchRuleDialog extends BaseDialog {
    private final AlopexFilter.Match match;
    private final DialogRuleBinding binding;
    private final List<Pair<AlopexFilter.Rule, String>> list;

    public MatchRuleDialog(Context context, AlopexFilter.Match match) {
        super(context, "Edit match rule");
        this.binding = DialogRuleBinding.inflate(LayoutInflater.from(context));
        setContent(binding.getRoot());
        this.match = match;
        this.binding.editName.setText(match.getName());
        this.list = new ArrayList<>();
        for (Pair<AlopexFilter.Rule, String> pair : match.getRuleList()) {
            this.list.add(pair.copy());
        }
        this.list.forEach(p -> addChip(this.binding.chipsRuleList, p, list));
        List<String> strings = new ArrayList<>();
        for (AlopexFilter.Rule rule : AlopexFilter.Rule.values()) {
            strings.add(rule.name());
        }
        this.binding.spinnerRuleMethod.setAdapter(new ArrayAdapter<>(context, android.R.layout.simple_spinner_dropdown_item, strings));
        this.binding.buttonAddNewRule.setOnClickListener(v -> {
            String pattern = this.binding.editNewRulePattern.getText().toString();
            if (pattern.isEmpty()) {
                AlopexView.showToast(context, "Enter pattern", Toast.LENGTH_SHORT);
                this.binding.editNewRulePattern.requestFocus();
                return;
            }
            this.binding.editNewRulePattern.setText("");
            Pair<AlopexFilter.Rule, String> pair = new Pair<>(AlopexFilter.Rule.valueOf(this.binding.spinnerRuleMethod.getSelectedItem().toString()), pattern);
            list.add(pair);
            addChip(binding.chipsRuleList, pair, list);
        });
    }

    @Override
    public void onConfirm() {
        String name = this.binding.editName.getText().toString();
        if (name.isEmpty()) {
            AlopexView.showToast(context, "Enter rule name", Toast.LENGTH_SHORT);
            this.binding.editName.requestFocus();
            return;
        }
        match.setName(name);
        match.setRuleList(list);
        super.onConfirm();
    }

    public void addChip(ChipGroup group, Pair<AlopexFilter.Rule, String> pair, List<Pair<AlopexFilter.Rule, String>> list) {
        Chip chip = new Chip(context);
        chip.setText(String.format("[%s]%s", pair.a.display(), pair.b));
        chip.setOnClickListener(v -> chip.setCloseIconVisible(!chip.isCloseIconVisible()));
        chip.setOnCloseIconClickListener(v ->
                new BaseDialog(context, "Delete rule?") {
                    @Override
                    public void onConfirm() {
                        for (Iterator<Pair<AlopexFilter.Rule, String>> iterator = list.listIterator(); iterator.hasNext(); ) {
                            Pair<AlopexFilter.Rule, String> m = iterator.next();
                            if (m == pair) {
                                iterator.remove();
                                group.removeView(chip);
                                break;
                            }
                        }
                        super.onConfirm();
                    }
                }.setButtonText("Yes", "No").show());
        group.addView(chip, new ChipGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));
    }
}

File: app/src/main/java/com/afoxxvi/alopex/ui/dialog/MatchRuleDialog.kt
package com.afoxxvi.alopex.ui.dialog

import android.R
import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Toast
import com.afoxxvi.alopex.Alopex
import com.afoxxvi.alopex.component.filter.AlopexFilter
import com.afoxxvi.alopex.component.filter.NotifyRule
import com.afoxxvi.alopex.databinding.DialogRuleBinding
import com.afoxxvi.alopex.util.Pair
import com.google.android.material.chip.Chip
import com.google.android.material.chip.ChipGroup
import java.util.function.Consumer

open class MatchRuleDialog(context: Context?, match: AlopexFilter.Match) : BaseDialog(context!!, "Edit match rule") {
    private val match: AlopexFilter.Match
    private val binding: DialogRuleBinding
    private val list: MutableList<Pair<NotifyRule, String>>

    init {
        binding = DialogRuleBinding.inflate(LayoutInflater.from(context))
        setContent(binding.root)
        this.match = match
        binding.editName.setText(match.name)
        list = ArrayList()
        match.ruleList.forEach { pair ->
            list.add(pair.copy())
        }
        list.forEach(Consumer { p: Pair<NotifyRule, String> -> addChip(binding.chipsRuleList, p, list) })
        val strings: MutableList<String> = ArrayList()
        for (rule in NotifyRule.values()) {
            strings.add(rule.name)
        }
        binding.spinnerRuleMethod.adapter = ArrayAdapter(context!!, R.layout.simple_spinner_dropdown_item, strings)
        binding.buttonAddNewRule.setOnClickListener {
            val pattern = binding.editNewRulePattern.text.toString()
            if (pattern.isEmpty()) {
                Alopex.showToast(context, "Enter pattern", Toast.LENGTH_SHORT)
                binding.editNewRulePattern.requestFocus()
                return@setOnClickListener
            }
            binding.editNewRulePattern.setText("")
            val pair = Pair(NotifyRule.valueOf(binding.spinnerRuleMethod.selectedItem.toString()), pattern)
            list.add(pair)
            addChip(binding.chipsRuleList, pair, list)
        }
    }

    override fun onConfirm() {
        val name = binding.editName.text.toString()
        if (name.isEmpty()) {
            Alopex.showToast(context, "Enter rule name", Toast.LENGTH_SHORT)
            binding.editName.requestFocus()
            return
        }
        match.name = name
        match.ruleList = list
        super.onConfirm()
    }

    private fun addChip(group: ChipGroup, pair: Pair<NotifyRule, String>, list: MutableList<Pair<NotifyRule, String>>) {
        val chip = Chip(context)
        chip.text = String.format("[%s]%s", pair.a.display(), pair.b)
        chip.setOnClickListener { chip.isCloseIconVisible = !chip.isCloseIconVisible }
        chip.setOnCloseIconClickListener {
            object : BaseDialog(context, "Delete rule?") {
                override fun onConfirm() {
                    list.remove(pair)
                    group.removeView(chip)
                    super.onConfirm()
                }
            }.setButtonText("Yes", "No").show()
        }
        group.addView(chip, ChipGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT))
    }
}